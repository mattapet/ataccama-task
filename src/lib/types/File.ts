/**
 * These types describe shape of the expected data.
 */

export interface Data {
  [key: string]: string;
}

export interface Kid {
  records: Record[];
}

export interface Kids {
  [key: string]: Kid;
}

export interface Record {
  _id: string;
  data: Data;
  kids: Kids;
}

export type Records = Record[];

/**
 * Root file representation.
 */
export type File = Record[];
