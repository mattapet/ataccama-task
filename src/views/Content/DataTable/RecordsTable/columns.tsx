//#region imports
import React from 'react';
import { ColumnProps } from 'antd/lib/table';
import { Record } from 'lib/types/File';
//#endregion

/**
 * Creates a column definition for given properties.
 *
 * Note that the function appends an `Action` column.
 *
 * @param properties An array of names of record properties
 * @param onRemove Handler invokation of which is supposed to remove record at
 * given index.
 * @returns A list of columns describing properties, if non-empty array of
 * `properties` is passed, otherwise an empty array.
 */
function columns(
  properties: string[],
  onRemove: (index: number) => any
): ColumnProps<Record>[] {
  if (!properties.length) {
    // If no properties are specified, return empty array to avoid having
    // a single (Action) column rendered.
    return [];
  }

  return [
    // Add columns for all common properties of records
    ...properties.map(property => ({
      dataIndex: `data.${property}`,
      title: property,
    })),
    // Append an `Actions` column with `Remove` button
    {
      title: 'Actions',
      render: (_, __, index) => (
        // eslint-disable-next-line
        <a href="javascript:;" onClick={() => onRemove(index)}>
          Remove
        </a>
      ),
    },
  ];
}

export default columns;
