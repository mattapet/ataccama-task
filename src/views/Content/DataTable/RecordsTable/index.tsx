//#region imports
import React, { useState } from 'react';
import { Record } from 'lib/types/File';
import View from './view';
//#endregion

//#region Component interfaces
export interface Props {
  /**
   * Title of the table -- used when nested table as name of the child table.
   */
  readonly title?: string;
  /**
   * List of records to display.
   */
  readonly records: Record[];
}
//#endregion

/**
 * Renders given `records` as a table with variable number of columns.
 *
 * The table contains columns, that are derived from the `records` passed to
 * the component as set of all properties of all of the records. For example
 * consider two records, one with properties `a, b, c` and the other with
 * properties `b, c, x`. The table rendered will would have columns `a, b, c`
 * and `x`.
 *
 * The `RecordsTable` is a statefull keeping track of records, which can be
 * removed.
 */
const RecordsTable: React.FunctionComponent<Props> = props => {
  // List of currently rendered items -- derived from given props
  const [records, setRecords] = useState(props.records);

  // Handler removing a record at specified `index`
  const onRemove = (index: number) => {
    const updatedRecords = records.filter((_, idx) => idx !== index);
    setRecords(updatedRecords);
  };

  return <View title={props.title} records={records} onRemove={onRemove} />;
};

export default RecordsTable;
