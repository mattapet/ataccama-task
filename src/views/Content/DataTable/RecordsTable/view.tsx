//#region imports
import React from 'react';
import { Record } from 'lib/types/File';
import { Table } from 'antd';
import columns from './columns';
import ExpandIcon from './components/ExpandIcon';
import KidsTable from '../KidsTable';
//#endregion

//#region Component interfaces
export interface Props {
  /**
   * Title of the table -- used when nested table as name of the child table.
   */
  readonly title?: string;
  /**
   * List of records to display.
   */
  readonly records: Record[];
  /**
   * Remove handler.
   *
   * @param index Index of the item that should be removed.
   */
  readonly onRemove: (index: number) => any;
}
//#endregion

const View: React.FunctionComponent<Props> = props => {
  const properties: Set<string> = props.records.reduce(
    (partial, next) => new Set([...partial, ...Object.keys(next.data)]),
    new Set()
  );

  return (
    <Table<Record>
      title={() => props.title && <h3>{props.title}</h3>}
      columns={columns([...properties], props.onRemove)}
      size="middle"
      dataSource={props.records}
      rowKey={record => record._id}
      expandIcon={ExpandIcon}
      expandedRowRender={record => <KidsTable data={record.kids} />}
      pagination={false}
    />
  );
};

export default View;
