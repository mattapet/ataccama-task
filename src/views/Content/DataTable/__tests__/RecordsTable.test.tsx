//#region imports
import React from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';

import RecordsTable from '../RecordsTable';
import { Record } from 'lib/types/File';
//#endregion

let container: HTMLDivElement | null = null;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  container && document.body.removeChild(container);
  container = null;
});

it('should render table with no columns, when no records provided', () => {
  // Arrange
  const records: Record[] = [];

  // Act
  act(() => {
    ReactDOM.render(<RecordsTable records={records} />, container);
  });

  // Assert
  const thead = container && container.querySelector('thead');
  expect(thead && thead.childNodes.length).toEqual(0);
});

it('should render table with right columns', () => {
  // Arrange
  const records: Record[] = [
    {
      _id: 't',
      data: { a: '2', b: '4' },
      kids: {},
    },
  ];

  // Act
  act(() => {
    ReactDOM.render(<RecordsTable records={records} />, container);
  });

  // Assert
  const thead = container && container.querySelector('thead');
  const tr = thead && thead.querySelector('tr');
  expect(tr && tr.childNodes.length).toEqual(4);
});

it('should render table with right columns', () => {
  // Arrange
  const records: Record[] = [
    {
      _id: 't',
      data: { a: '2', b: '4' },
      kids: {},
    },
    {
      _id: 'f',
      data: { a: '2', c: '4' },
      kids: {},
    },
  ];

  // Act
  act(() => {
    ReactDOM.render(<RecordsTable records={records} />, container);
  });

  // Assert
  const thead = container && container.querySelector('thead');
  const tr = thead && thead.querySelector('tr');
  expect(tr && tr.childNodes.length).toEqual(5);

  const tbody = container && container.querySelector('tbody');
  expect(tbody && tbody.childNodes.length).toEqual(records.length);
});

it('should remove record when `remove` button clicked', () => {
  // Arrange
  const records: Record[] = [
    {
      _id: 't',
      data: { a: '2', b: '4' },
      kids: {},
    },
    {
      _id: 'f',
      data: { a: '2', c: '4' },
      kids: {},
    },
  ];
  act(() => {
    ReactDOM.render(<RecordsTable records={records} />, container);
  });

  // Act
  const a = container && container.querySelector('a');
  act(() => {
    a && a.dispatchEvent(new MouseEvent('click', { bubbles: true }));
  });

  // Assert
  const tbody = container && container.querySelector('tbody');
  expect(tbody && tbody.childNodes.length).toEqual(records.length - 1);
});

it('should expand to kids table', () => {
  // Arrange
  const records: Record[] = [
    {
      _id: 't',
      data: { a: '2', b: '4' },
      kids: {
        test: {
          records: [
            {
              _id: 'f',
              data: { a: '2', c: '4' },
              kids: {},
            },
          ],
        },
      },
    },
  ];
  act(() => {
    ReactDOM.render(<RecordsTable records={records} />, container);
  });

  // Act
  const tbody0 = container && container.querySelector('tbody');
  const i = tbody0 && tbody0.querySelector('i');
  act(() => {
    i && i.dispatchEvent(new MouseEvent('click', { bubbles: true }));
  });

  // Assert
  const h3 = tbody0 && tbody0.querySelector('h3');
  const tbody1 = tbody0 && tbody0.querySelector('tbody');
  expect(h3 && h3.textContent).toEqual('test');
  expect(tbody1).toBeTruthy();
});
