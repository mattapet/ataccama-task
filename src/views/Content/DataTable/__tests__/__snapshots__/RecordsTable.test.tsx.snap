// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="ant-table-wrapper"
>
  <div
    className="ant-spin-nested-loading"
  >
    <div
      className="ant-spin-container"
    >
      <div
        className="ant-table ant-table-default ant-table-scroll-position-left"
        style={Object {}}
      >
        <div
          className="ant-table-title"
        />
        <div
          className="ant-table-content"
        >
          <div
            className="ant-table-body"
            onScroll={[Function]}
            onWheel={[Function]}
            style={Object {}}
          >
            <table
              className=""
              style={Object {}}
            >
              <colgroup>
                <col
                  className="ant-table-expand-icon-col"
                />
                <col
                  style={
                    Object {
                      "minWidth": undefined,
                      "width": undefined,
                    }
                  }
                />
                <col
                  style={
                    Object {
                      "minWidth": undefined,
                      "width": undefined,
                    }
                  }
                />
                <col
                  style={
                    Object {
                      "minWidth": undefined,
                      "width": undefined,
                    }
                  }
                />
              </colgroup>
              <thead
                className="ant-table-thead"
              >
                <tr
                  style={
                    Object {
                      "height": null,
                    }
                  }
                >
                  <th
                    className="ant-table-expand-icon-th"
                    rowSpan={1}
                    title=""
                  />
                  <th
                    className=""
                  >
                    <span
                      className="ant-table-header-column"
                    >
                      <div>
                        <span
                          className="ant-table-column-title"
                        >
                          a
                        </span>
                        <span
                          className="ant-table-column-sorter"
                        />
                      </div>
                    </span>
                  </th>
                  <th
                    className=""
                  >
                    <span
                      className="ant-table-header-column"
                    >
                      <div>
                        <span
                          className="ant-table-column-title"
                        >
                          b
                        </span>
                        <span
                          className="ant-table-column-sorter"
                        />
                      </div>
                    </span>
                  </th>
                  <th
                    className=""
                  >
                    <span
                      className="ant-table-header-column"
                    >
                      <div>
                        <span
                          className="ant-table-column-title"
                        >
                          Actions
                        </span>
                        <span
                          className="ant-table-column-sorter"
                        />
                      </div>
                    </span>
                  </th>
                </tr>
              </thead>
              <tbody
                className="ant-table-tbody"
              >
                <tr
                  className="ant-table-row ant-table-row-level-0"
                  data-row-key="t"
                  onClick={[Function]}
                  onContextMenu={[Function]}
                  onDoubleClick={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  style={
                    Object {
                      "height": null,
                    }
                  }
                >
                  <td
                    className="ant-table-row-expand-icon-cell"
                  />
                  <td
                    className=""
                    onClick={[Function]}
                  >
                    2
                  </td>
                  <td
                    className=""
                    onClick={[Function]}
                  >
                    4
                  </td>
                  <td
                    className=""
                    onClick={[Function]}
                  >
                    <a
                      href="javascript:;"
                      onClick={[Function]}
                    >
                      Remove
                    </a>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render correctly 2`] = `
<div
  className="ant-table-wrapper"
>
  <div
    className="ant-spin-nested-loading"
  >
    <div
      className="ant-spin-container"
    >
      <div
        className="ant-table ant-table-default ant-table-scroll-position-left"
        style={Object {}}
      >
        <div
          className="ant-table-title"
        />
        <div
          className="ant-table-content"
        >
          <div
            className="ant-table-body"
            onScroll={[Function]}
            onWheel={[Function]}
            style={Object {}}
          >
            <table
              className=""
              style={Object {}}
            >
              <colgroup>
                <col
                  className="ant-table-expand-icon-col"
                />
                <col
                  style={
                    Object {
                      "minWidth": undefined,
                      "width": undefined,
                    }
                  }
                />
                <col
                  style={
                    Object {
                      "minWidth": undefined,
                      "width": undefined,
                    }
                  }
                />
                <col
                  style={
                    Object {
                      "minWidth": undefined,
                      "width": undefined,
                    }
                  }
                />
              </colgroup>
              <thead
                className="ant-table-thead"
              >
                <tr
                  style={
                    Object {
                      "height": null,
                    }
                  }
                >
                  <th
                    className="ant-table-expand-icon-th"
                    rowSpan={1}
                    title=""
                  />
                  <th
                    className=""
                  >
                    <span
                      className="ant-table-header-column"
                    >
                      <div>
                        <span
                          className="ant-table-column-title"
                        >
                          a
                        </span>
                        <span
                          className="ant-table-column-sorter"
                        />
                      </div>
                    </span>
                  </th>
                  <th
                    className=""
                  >
                    <span
                      className="ant-table-header-column"
                    >
                      <div>
                        <span
                          className="ant-table-column-title"
                        >
                          b
                        </span>
                        <span
                          className="ant-table-column-sorter"
                        />
                      </div>
                    </span>
                  </th>
                  <th
                    className=""
                  >
                    <span
                      className="ant-table-header-column"
                    >
                      <div>
                        <span
                          className="ant-table-column-title"
                        >
                          Actions
                        </span>
                        <span
                          className="ant-table-column-sorter"
                        />
                      </div>
                    </span>
                  </th>
                </tr>
              </thead>
              <tbody
                className="ant-table-tbody"
              >
                <tr
                  className="ant-table-row ant-table-row-level-0"
                  data-row-key="t"
                  onClick={[Function]}
                  onContextMenu={[Function]}
                  onDoubleClick={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  style={
                    Object {
                      "height": null,
                    }
                  }
                >
                  <td
                    className="ant-table-row-expand-icon-cell"
                  >
                    <i
                      aria-label="icon: right"
                      className="anticon anticon-right"
                      onClick={[Function]}
                      tabIndex={-1}
                    >
                      <svg
                        aria-hidden="true"
                        className=""
                        data-icon="right"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z"
                        />
                      </svg>
                    </i>
                  </td>
                  <td
                    className=""
                    onClick={[Function]}
                  >
                    2
                  </td>
                  <td
                    className=""
                    onClick={[Function]}
                  >
                    4
                  </td>
                  <td
                    className=""
                    onClick={[Function]}
                  >
                    <a
                      href="javascript:;"
                      onClick={[Function]}
                    >
                      Remove
                    </a>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
